---
title: no-misleading-character-class - Rules
layout: doc
edit_link: https://github.com/eslint/eslint/edit/master/docs/rules/no-misleading-character-class.md
rule_type: problem
---
<!-- Note: No pull requests accepted for this file. See README.md in the root directory for details. -->

# Disallow characters which are made with multiple code points in character class syntax (no-misleading-character-class)

# ä¸å…è®¸åœ¨å­—ç¬¦ç±»è¯­æ³•ä¸­å‡ºç°ç”±å¤šä¸ªä»£ç ç‚¹ç»„æˆçš„å­—ç¬¦ (no-misleading-character-class)

Unicode includes the characters which are made with multiple code points.
RegExp character class syntax (`/[abc]/`) cannot handle characters which are made by multiple code points as a character; those characters will be dissolved to each code point. For example, `â‡ï¸` is made by `â‡` (`U+2747`) and VARIATION SELECTOR-16 (`U+FE0F`). If this character is in RegExp character class, it will match to either `â‡` (`U+2747`) or VARIATION SELECTOR-16 (`U+FE0F`) rather than `â‡ï¸`.

Unicode åŒ…æ‹¬ç”±å¤šä¸ªä»£ç ç‚¹ç»„æˆçš„å­—ç¬¦ã€‚RegExp å­—ç¬¦ç±»è¯­æ³• (`/[abc]/`) ä¸èƒ½å¤„ç†ç”±å¤šä¸ªä»£ç ç‚¹ç»„æˆçš„å­—ç¬¦ï¼›è¿™äº›å­—ç¬¦å°†è¢«åˆ†è§£åˆ°æ¯ä¸ªä»£ç ç‚¹ã€‚ä¾‹å¦‚ï¼Œ`â‡ï¸` æ˜¯ç”± `â‡` (`U+2747`) å’Œ VARIATION SELECTOR-16 (`U+FE0F`)ã€‚å¦‚æœè¿™æ˜¯æ­£åˆ™è¡¨è¾¾å¼å­—ç¬¦ç±»ï¼Œå®ƒå°†åŒ¹é… `â‡` (`U+2747`) æˆ– VARIATION SELECTOR-16 (`U+FE0F`)è€Œä¸æ˜¯ `â‡ï¸`ã€‚

This rule reports the regular expressions which include multiple code point characters in character class syntax. This rule considers the following characters as multiple code point characters.

æ­¤è§„åˆ™æŠ¥å‘Šåœ¨å­—ç¬¦ç±»è¯­æ³•ä¸­åŒ…å«å¤šä¸ªä»£ç ç‚¹å­—ç¬¦çš„æ­£åˆ™è¡¨è¾¾å¼ã€‚æ­¤è§„åˆ™å°†ä»¥ä¸‹å­—ç¬¦è§†ä¸ºå¤šä¸ªä»£ç ç‚¹å­—ç¬¦ã€‚

**A character with combining characters:**

**ç»„åˆå­—ç¬¦çš„å­—ç¬¦ï¼š**

The combining characters are characters which belong to one of `Mc`, `Me`, and `Mn` [Unicode general categories](http://www.unicode.org/L2/L1999/UnicodeData.html#General%20Category).

ç»„åˆå­—ç¬¦å±äº `Mc`ã€`Me` å’Œ `Mn` [Unicode é€šç”¨ç±»åˆ«](http://www.unicode.org/L2/L1999/UnicodeData.html#General%20Category) ä¹‹ä¸€ã€‚

```js
/^[AÌ]$/u.test("AÌ") //â†’ false
/^[â‡ï¸]$/u.test("â‡ï¸") //â†’ false
```

**A character with Emoji modifiers:**

**å¸¦æœ‰è¡¨æƒ…ç¬¦å·ä¿®é¥°ç¬¦çš„å­—ç¬¦ï¼š**

```js
/^[ğŸ‘¶ğŸ»]$/u.test("ğŸ‘¶ğŸ»") //â†’ false
/^[ğŸ‘¶ğŸ½]$/u.test("ğŸ‘¶ğŸ½") //â†’ false
```

**A pair of regional indicator symbols:**

**ä¸€ç»„åŒºåŸŸæŒ‡æ ‡ç¬¦å·ï¼š**

```js
/^[ğŸ‡¯ğŸ‡µ]$/u.test("ğŸ‡¯ğŸ‡µ") //â†’ false
```

**Characters that ZWJ joins:**

**ZWJï¼ˆZero Width Joinerï¼‰è¿æ¥çš„å­—ç¬¦ï¼š**

```js
/^[ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦]$/u.test("ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦") //â†’ false
```

**A surrogate pair without Unicode flag:**

**æ²¡æœ‰ Unicode æ ‡å¿—çš„ Surrogate pairï¼š**

```js
/^[ğŸ‘]$/.test("ğŸ‘") //â†’ false

// Surrogate pair is OK if with u flag.
/^[ğŸ‘]$/u.test("ğŸ‘") //â†’ true
```

## Rule Details

This rule reports the regular expressions which include multiple code point characters in character class syntax.

æ­¤è§„åˆ™æŠ¥å‘Šåœ¨å­—ç¬¦ç±»è¯­æ³•ä¸­åŒ…å«å¤šä¸ªä»£ç ç‚¹å­—ç¬¦çš„æ­£åˆ™è¡¨è¾¾å¼ã€‚

Examples of **incorrect** code for this rule:

**é”™è¯¯** ä»£ç ç¤ºä¾‹ï¼š

```js
/*eslint no-misleading-character-class: error */

/^[AÌ]$/u
/^[â‡ï¸]$/u
/^[ğŸ‘¶ğŸ»]$/u
/^[ğŸ‡¯ğŸ‡µ]$/u
/^[ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦]$/u
/^[ğŸ‘]$/
```

Examples of **correct** code for this rule:

**æ­£ç¡®** ä»£ç ç¤ºä¾‹ï¼š

```js
/*eslint no-misleading-character-class: error */

/^[abc]$/
/^[ğŸ‘]$/u
```

## When Not To Use It

You can turn this rule off if you don't want to check RegExp character class syntax for multiple code point characters.

å¦‚æœä¸æƒ³æ£€æŸ¥å¤šä¸ªä»£ç ç‚¹å­—ç¬¦çš„ RegExp å­—ç¬¦ç±»è¯­æ³•ï¼Œå¯ä»¥å…³é—­æ­¤è§„åˆ™ã€‚

## Version

This rule was introduced in ESLint 5.3.0.

è¯¥è§„åˆ™åœ¨ ESLint 5.3.0 è¢«å¼•å…¥ã€‚

## Resources

* [Rule source](https://github.com/eslint/eslint/tree/master/lib/rules/no-misleading-character-class.js)
* [Documentation source](https://github.com/eslint/eslint/tree/master/docs/rules/no-misleading-character-class.md)
