---
title: require-unicode-regexp - Rules
layout: doc
edit_link: https://github.com/eslint/eslint/edit/master/docs/rules/require-unicode-regexp.md
rule_type: suggestion
---
<!-- Note: No pull requests accepted for this file. See README.md in the root directory for details. -->

# Enforce the use of `u` flag on RegExp (require-unicode-regexp)
# å¼ºåˆ¶åœ¨ RegExp ä¸Šä½¿ç”¨ `u` æ ‡å¿— (require-unicode-regexp)

RegExp `u` flag has two effects:

RegExp çš„ `u` æ ‡å¿—æœ‰ä¸¤ä¸ªæ•ˆæœ:

1. **Make the regular expression handling UTF-16 surrogate pairs correctly.**
1. **ä½¿æ­£åˆ™è¡¨è¾¾å¼æ­£ç¡®å¤„ç† UTF-16 Surrogate Pair**

Especially, character range syntax gets the correct behavior.

ç‰¹åˆ«æ˜¯ï¼Œå­—ç¬¦èŒƒå›´è¯­æ³•è·å¾—æ­£ç¡®çš„è¡Œä¸ºã€‚

    ```js
    /^[ğŸ‘]$/.test("ğŸ‘") //â†’ false
    /^[ğŸ‘]$/u.test("ğŸ‘") //â†’ true
    ```

2. **Make the regular expression throwing syntax errors early as disabling [Annex B extensions](https://www.ecma-international.org/ecma-262/6.0/#sec-regular-expressions-patterns).**
2. **ä½¿æ­£åˆ™è¡¨è¾¾å¼åœ¨ç¦ç”¨æ—¶å°½æ—©æŠ›å‡ºè¯­æ³•é”™è¯¯ [Annex B extensions](https://www.ecma-international.org/ecma-262/6.0/#sec-regular-expressions-patterns)**

Because of historical reason, JavaScript regular expressions are tolerant of syntax errors. For example, `/\w{1, 2/` is a syntax error, but JavaScript doesn't throw the error. It matches strings such as `"a{1, 2"` instead. Such a recovering logic is defined in Annex B.

ç”±äºå†å²åŸå› ï¼ŒJavaScript æ­£åˆ™è¡¨è¾¾å¼èƒ½å¤Ÿå®¹å¿è¯­æ³•é”™è¯¯ã€‚ä¾‹å¦‚ï¼Œ`/\w{1, 2/` æ˜¯ä¸€ä¸ªè¯­æ³•é”™è¯¯ï¼Œä½†æ˜¯JavaScriptä¸ä¼šæŠ›å‡ºè¿™ä¸ªé”™è¯¯ã€‚å®ƒä¼šåŒ¹é…å­—ç¬¦ä¸²ï¼Œæ¯”å¦‚ `"a{1, 2"`ã€‚è¿™ç§æ¢å¤é€»è¾‘åœ¨é™„ä»¶Bä¸­å®šä¹‰ã€‚

The `u` flag disables the recovering logic Annex B defined. As a result, you can find errors early. This is similar to [the strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode).

`u` æ ‡å¿—ç¦ç”¨åœ¨é™„ä»¶Bä¸­å®šä¹‰çš„æ¢å¤é€»è¾‘ã€‚å› æ­¤ï¼Œä½ å¯ä»¥å°½æ—©å‘ç°é”™è¯¯ã€‚è¿™ç±»ä¼¼äº[ä¸¥æ ¼æ¨¡å¼](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode)

Therefore, the `u` flag lets us work better with regular expressions.

å› æ­¤ï¼Œ`u` æ ‡å¿—ä½¿æˆ‘ä»¬èƒ½å¤Ÿæ›´å¥½åœ°å¤„ç†æ­£åˆ™è¡¨è¾¾å¼ã€‚

## Rule Details

This rule aims to enforce the use of `u` flag on regular expressions.

è¯¥è§„åˆ™æ—¨åœ¨å¼ºåˆ¶æ­£åˆ™è¡¨è¾¾å¼ä½¿ç”¨ `u` æ ‡å¿—ã€‚

Examples of **incorrect** code for this rule:

**é”™è¯¯** ä»£ç ç¤ºä¾‹ï¼š

```js
/*eslint require-unicode-regexp: error */

const a = /aaa/
const b = /bbb/gi
const c = new RegExp("ccc")
const d = new RegExp("ddd", "gi")
```

Examples of **correct** code for this rule:

**æ­£ç¡®** ä»£ç ç¤ºä¾‹ï¼š

```js
/*eslint require-unicode-regexp: error */

const a = /aaa/u
const b = /bbb/giu
const c = new RegExp("ccc", "u")
const d = new RegExp("ddd", "giu")

// This rule ignores RegExp calls if the flags could not be evaluated to a static value.
function f(flags) {
    return new RegExp("eee", flags)
}
```

## When Not To Use It

If you don't want to notify regular expressions with no `u` flag, then it's safe to disable this rule.

å¦‚æœä¸å¸Œæœ›é€šçŸ¥æ²¡æœ‰ `u` æ ‡å¿—çš„æ­£åˆ™è¡¨è¾¾å¼ï¼Œåˆ™å¯ä»¥ç¦ç”¨æ­¤è§„åˆ™ã€‚

## Version

This rule was introduced in ESLint 5.3.0.

è¯¥è§„åˆ™åœ¨ ESLint 5.3.0 ä¸­è¢«å¼•å…¥ã€‚

## Resources

* [Rule source](https://github.com/eslint/eslint/tree/master/lib/rules/require-unicode-regexp.js)
* [Documentation source](https://github.com/eslint/eslint/tree/master/docs/rules/require-unicode-regexp.md)
